cmd/utils.go

```diff
@@ -970,3 +971,164 @@ func auditLogInternal(ctx context.Context, bucket, object string, opts AuditLogO
 	ctx = logger.SetAuditEntry(ctx, &entry)
 	logger.AuditLog(ctx, nil, nil, nil)
 }
+
+type speedtestResults []SpeedtestResult
+
+func (s speedtestResults) Len() int {
+	return len(s)
+}
+
+func (s speedtestResults) Less(i, j int) bool {
+	return s[i].Endpoint < s[i].Endpoint
+}
+
+func (s speedtestResults) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+// Get the max throughput and iops numbers.
+func speedTest(ctxt context.Context, w io.Writer, throughputSize, iopsSize int, concurrencyStart int, duration time.Duration, autotune bool) (madmin.SpeedTestResult, error) {
+	var result madmin.SpeedTestResult
+
+	objAPI := newObjectLayerFn()
+	if objAPI == nil {
+		return result, errServerNotInitialized
+	}
+
+	concurrency := concurrencyStart
+
+	throughputHighestGet := uint64(0)
+	throughputHighestPut := uint64(0)
+	var throughputHighestPutResults []SpeedtestResult
+	var throughputHighestGetResults []SpeedtestResult
+
+	for {
+		// Write a blank entry to prevent client from disconnecting
+		if err := json.NewEncoder(w).Encode(madmin.SpeedTestResult{}); err != nil {
+			return result, err
+		}
+		w.(http.Flusher).Flush()
+		results := globalNotificationSys.Speedtest(ctxt, throughputSize, concurrency, duration)
+		sort.Sort(speedtestResults(results))
+		totalPut := uint64(0)
+		totalGet := uint64(0)
+		for _, result := range results {
+			totalPut += result.Uploads
+			totalGet += result.Downloads
+		}
+		if totalPut < throughputHighestPut && totalGet < throughputHighestGet {
+			break
+		}
+
+		if totalPut > throughputHighestPut {
+			throughputHighestPut = totalPut
+			throughputHighestPutResults = results
+		}
+		if totalGet > throughputHighestGet {
+			throughputHighestGet = totalGet
+			throughputHighestGetResults = results
+		}
+		if !autotune {
+			break
+		}
+		// Try with a higher concurrency to see if we get better throughput
+		concurrency += 8
+	}
+
+	concurrency = concurrencyStart
+	iopsHighestPut := uint64(0)
+	iopsHighestGet := uint64(0)
+	var iopsHighestPutResults []SpeedtestResult
+	var iopsHighestGetResults []SpeedtestResult
+
+	if autotune {
+		for {
+			// Write a blank entry to prevent client from disconnecting
+			if err := json.NewEncoder(w).Encode(madmin.SpeedTestResult{}); err != nil {
+				return result, err
+			}
+			w.(http.Flusher).Flush()
+			results := globalNotificationSys.Speedtest(ctxt, iopsSize, concurrency, duration)
+			sort.Sort(speedtestResults(results))
```

Comment:
```suggestion
			sort.Slice(results, func(i, j int) bool {
				return results[i].Endpoint < results[j].Endpoint
			})
```
Remove the `speedtestResults` type.